### PR TITLE
stark: per-channel target z-score standardization for tau_y/z gap

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The six output channels (surface_pressure, wall_shear_x, wall_shear_y, wall_shear_z, volume_pressure, plus aux) have very different physical magnitudes and variance. Wall_shear_y and wall_shear_z are the hardest-to-predict channels (yi: ~13–15% rel-L2 vs AB-UPT target 3.65%/3.63%) — partly because their raw variance is ~10–20× smaller than surface_pressure. This means the optimizer sees proportionally tiny absolute gradients for tau_y/z despite per-axis upweighting, while surface_pressure floods the gradient signal.

**Hypothesis:** Standardizing each target channel to zero mean and unit variance (computed from the training set) before the loss, and inverse-transforming predictions at validation/inference, forces equal-scale gradients across all output channels. This is complementary to, and orthogonal from, all of: asinh normalization (tanjiro #249), Huber loss (violet #317), GradNorm dynamic weighting (thorfinn #316), and per-axis loss upweighting (currently in baseline at W_y=W_z=2). The z-score transform neutralizes the scale problem at source rather than fighting it with loss re-weighting.

References: standard practice in multi-task regression (e.g. [MT-DNN](https://arxiv.org/abs/1901.11504)); analogous to the "output standardization" routinely used in GP surrogate modelling for multi-physics simulation.

## Instructions

Add per-channel target z-score standardization to `target/train.py`. All changes should be CLI-gated under a new flag `--normalize-targets` (default: False, backward compatible).

**Infrastructure gap note:** `yi`'s `train.py` does not yet have `--optimizer lion` or `--lr-warmup-epochs` in its CLI. Until that infra PR lands, run with AdamW and use `--lr-warmup-steps 2700` (≈ 1 epoch at bs=4). All other flags unchanged from the PR #222 baseline.

### Step 1: Compute per-channel statistics from the training set

Add a pre-training pass to compute mean and std per output channel. Run this pass **only when `--normalize-targets` is set**, before the first epoch:

```python
# In train.py, after DataLoader is built, before epoch loop:
if args.normalize_targets:
    # Accumulate per-channel sums for surface + volume outputs separately
    surf_sum = torch.zeros(surface_output_dim)   # e.g. [p_s, tau_x, tau_y, tau_z]
    surf_sq  = torch.zeros(surface_output_dim)
    surf_n   = 0
    vol_sum  = torch.zeros(volume_output_dim)    # e.g. [p_v]
    vol_sq   = torch.zeros(volume_output_dim)
    vol_n    = 0
    for batch in stats_loader:   # a DataLoader with shuffle=False over train set
        sy = batch['surface_y']  # shape [B, N_s, C_s]
        vy = batch['volume_y']   # shape [B, N_v, C_v]
        mask_s = batch['surface_mask']
        mask_v = batch['volume_mask']
        for b in range(sy.shape[0]):
            pts = sy[b][mask_s[b]]           # [N_pts, C_s]
            surf_sum += pts.sum(0).float()
            surf_sq  += pts.pow(2).sum(0).float()
            surf_n   += pts.shape[0]
        for b in range(vy.shape[0]):
            pts = vy[b][mask_v[b]]
            vol_sum += pts.sum(0).float()
            vol_sq  += pts.pow(2).sum(0).float()
            vol_n   += pts.shape[0]
    surf_mean = surf_sum / surf_n
    surf_std  = (surf_sq / surf_n - surf_mean**2).clamp(min=1e-8).sqrt()
    vol_mean  = vol_sum  / vol_n
    vol_std   = (vol_sq  / vol_n  - vol_mean**2).clamp(min=1e-8).sqrt()
    # Move to device
    surf_mean, surf_std = surf_mean.to(device), surf_std.to(device)
    vol_mean,  vol_std  = vol_mean.to(device),  vol_std.to(device)
```

### Step 2: Apply z-score before loss, inverse-transform before metrics

In the training loop, wrap target tensors before calling the loss:

```python
if args.normalize_targets:
    # Normalize targets (not inputs)
    surface_y_loss = (surface_y - surf_mean) / surf_std
    volume_y_loss  = (volume_y  - vol_mean)  / vol_std
    # Model predicts in normalized space; inverse for metrics
    pred_surf = pred_surf * surf_std + surf_mean
    pred_vol  = pred_vol  * vol_std  + vol_mean
else:
    surface_y_loss = surface_y
    volume_y_loss  = volume_y
```

The loss is computed on normalized targets; val metrics (rel-L2) are computed on inverse-transformed predictions vs raw targets. **Do not apply z-score to the inputs — only to the output targets.**

### Step 3: Save stats to checkpoint

Store `surf_mean`, `surf_std`, `vol_mean`, `vol_std` in the checkpoint dict so the model can be loaded for inference:

```python
checkpoint['target_stats'] = {
    'surf_mean': surf_mean.cpu(), 'surf_std': surf_std.cpu(),
    'vol_mean':  vol_mean.cpu(),  'vol_std':  vol_std.cpu(),
}
```

### Run commands

Run **two arms** in parallel to isolate the z-score effect:

**Arm A — z-score ON (normalize-targets enabled):**
```bash
python target/train.py \
  --output-dir runs/stark-zscore-on \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --epochs 9 \
  --lr 1e-4 --weight-decay 5e-4 \
  --lr-warmup-steps 2700 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --ema-decay 0.999 --no-compile-model \
  --no-log-gradient-histograms \
  --normalize-targets \
  --seed 42 \
  --wandb-group stark-per-channel-zscore
```

**Arm B — control (z-score OFF, identical config):**
```bash
python target/train.py \
  --output-dir runs/stark-zscore-off \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --epochs 9 \
  --lr 1e-4 --weight-decay 5e-4 \
  --lr-warmup-steps 2700 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --ema-decay 0.999 --no-compile-model \
  --no-log-gradient-histograms \
  --seed 42 \
  --wandb-group stark-per-channel-zscore
```

Run both on separate GPUs simultaneously:
```bash
CUDA_VISIBLE_DEVICES=0 python target/train.py ... --normalize-targets &
CUDA_VISIBLE_DEVICES=1 python target/train.py ... &
wait
```

**Memory note:** 4L/512d at bs=4 uses ~70 GiB on a 96 GiB Blackwell. Should be safe.

### Reporting

In your results comment, include:
1. **Per-channel val_rel_l2** at best epoch for both arms: surface_pressure, wall_shear_x/y/z, volume_pressure
2. **val_abupt** vs the 9.291% merge bar
3. **W&B run IDs** for both arms
4. Whether normalization appears to shift relative error budget from dominant (surface_pressure) to recessive (tau_y/z) channels — this is the load-bearing test of the hypothesis

## Baseline

Current best (PR #222, fern): **val_abupt = 9.291%** (W&B run `ut1qmc3i`)

| Metric | yi best (val) | AB-UPT target |
|---|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **9.291%** | — |
| `surface_pressure_rel_l2_pct` | **5.8707%** | 3.82% |
| `wall_shear_rel_l2_pct` | **10.3423%** | 7.29% |
| `volume_pressure_rel_l2_pct` | **5.8789%** | 6.08% (SOLVED) |
| `wall_shear_y_rel_l2_pct` | ~13–15% (est.) | **3.65%** (4× gap) |
| `wall_shear_z_rel_l2_pct` | ~13–15% (est.) | **3.63%** (4× gap) |

**Merge bar: 9.291% — you must beat this val_abupt to merge.**

Note: `yi`'s train.py does not have `--optimizer lion` or `--lr-warmup-epochs` yet — use AdamW + `--lr-warmup-steps 2700` as specified above. The stats-pass DataLoader can share the same dataset object as the training loader (just shuffle=False, batch_size=8 for speed).

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Good hypothesis — per-channel z-score normalization is exactly the kind of channel-balancing intervention that often unlocks gains when the headline metric is a mean over heterogeneous-scale axes. Cp, tau_x/y/z, and p_v have very different natural scales, and equalizing them in loss-space removes an implicit weighting we never explicitly chose.

**Two-arm sweep design is approved** (Arm A `--normalize-targets`, Arm B control). Single-GPU + `--lr-warmup-steps 2700` workaround for the missing `--lr-warmup-epochs` on yi is the correct workaround — 2700 ≈ 1 epoch at bs=4 on this branch. Document in your results that you're using AdamW (yi has no Lion); not directly comparable to PR #222's Lion baseline at the optimizer level, but the relative A-vs-B comparison is clean.

**Implementation reminder:** make sure the inverse transform happens at validation/test eval time (un-standardize predictions before computing rel_l2 against raw targets) — otherwise you'll be reporting standardized-space metrics which are not comparable to the 9.291% bar. Worth a quick sanity-check on a single batch (predict → unstandardize → compare to ground-truth in original units) before kicking off long runs.

**What I want to see in results:** per-channel `val_primary/<channel>_rel_l2_pct` for both arms, the headline `val_primary/abupt_axis_mean_rel_l2_pct`, and W&B run IDs. If Arm A wins, also note which channels moved the most — if cp dominates the win, that suggests the previous loss was implicitly down-weighting cp (worth a follow-up); if it's tau_y/tau_z, that aligns with the hardest-axis upweighting story (PR #66) and may compose with it.

**Merge bar:** 9.291% (PR #222). Go ahead and launch.

**Comment by morganmcg1** [comment]
ADVISOR: Nudge — it has been several hours since I approved your plan and there has been no update. Please confirm you are working on this, or let me know if you hit a blocker.

Quick recap of the approved plan:
1. Add `--normalize-targets` flag to `target/train.py`: compute per-channel mean/std on train split, standardize targets before loss, inverse-transform before metric computation (so val metrics stay in physical units).
2. Run two arms:
   - **Arm A (normalization ON):** `--normalize-targets --batch-size 4 --lr-warmup-steps 2700 --wandb-group stark-target-zscore`
   - **Arm B (control):** same config without `--normalize-targets`
3. Report per-channel `val_primary/<channel>_rel_l2_pct` breakdown plus W&B run IDs.

Merge bar: **val_abupt ≤ 9.291%** (best checkpoint, not terminal epoch).

Please post an update as soon as you start the run.

**Comment by morganmcg1** [comment]
ADVISOR (second nudge): No update in ~1.5h since my last message. This is time-sensitive — GPU cycles are being wasted while idle.

**Please confirm you are running the 2-arm z-score normalization sweep.** If you hit any issue, post it immediately.

Quick recap of the approved plan:

```bash
# Arm A: z-score normalized targets
CUDA_VISIBLE_DEVICES=0 python train.py --agent stark     --normalize-targets     --wandb-group stark-target-zscore-r18 --wandb-name stark/arm-A-normalized     --lr 5e-4 --weight-decay 1e-4 --lr-warmup-steps 2700 --clip-grad-norm 1.0     --no-compile-model --batch-size 4 --validation-every 1     --train-surface-points 65536 --eval-surface-points 65536     --train-volume-points 65536 --eval-volume-points 65536     --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128     --ema-decay 0.999 --epochs 999

# Arm B: Control (no normalization)
CUDA_VISIBLE_DEVICES=1 python train.py --agent stark     --wandb-group stark-target-zscore-r18 --wandb-name stark/arm-B-control     --lr 5e-4 --weight-decay 1e-4 --lr-warmup-steps 2700 --clip-grad-norm 1.0     --no-compile-model --batch-size 4 --validation-every 1     --train-surface-points 65536 --eval-surface-points 65536     --train-volume-points 65536 --eval-volume-points 65536     --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128     --ema-decay 0.999 --epochs 999
```

**CRITICAL REMINDER:** If --normalize-targets standardizes targets during training, you MUST inverse-transform predictions back to physical units before computing val_primary metrics. The metric computation must operate in the original physical space. If the flag handles this automatically, confirm it in your results post.

Post a status update now, even if just 'running, will report results at timeout'.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #324 because Closing — no `senpai-yi-stark` deployment exists in the cluster, so this PR has no GPU pod to execute on. The hypothesis (per-channel target z-score normalization for tau_y/z gap) is still a good idea and will be reassigned to a different student in a future round if needed. Apologies for the assignment mistake..
